src/MongoDB.Driver.Core/Core/Misc/StreamExtensionMethods.cs

```diff
@@ -116,6 +149,37 @@ public static async Task ReadBytesAsync(this Stream stream, IByteBuffer buffer,
             }
         }
 
+
+        public static async Task WriteAsync(this Stream stream, byte[] buffer, int offset, int count, TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            var state = 1; // 1 == writing, 2 == done writing, 3 == timedout, 4 == cancelled
+            void changeState(int to)
+            {
+                var from = Interlocked.CompareExchange(ref state, to, 1);
+                if (from == 1 && to >= 3)
+                {
+                    try { stream.Dispose(); } catch { } // disposing the stream aborts the write attempt
+                }
+            }
+
+            using (new Timer(_ => changeState(3), null, timeout, Timeout.InfiniteTimeSpan))
+            using (cancellationToken.Register(() => changeState(4)))
+            {
+                try
+                {
+                    await stream.WriteAsync(buffer, offset, count, cancellationToken).ConfigureAwait(false);
+                    changeState(2);
```

Comment:
It looks like there is a very small possibility that line 170 will be successfully executed, but right after this line (and before 171) we will catch a timeout and then dispose the stream. So, we will see the exception even though the data was successfully sent to the server. 

But, as I remember we discussed this case before and decided to leave it as is, so, maybe it's ok to not change this part.

Reply:
The same comment as above, if I'm not mistaken the state `2` is not necessary here, it's enough just to call:

    return await stream.WriteAsync(buffer, offset, count, cancellationToken).ConfigureAwait(false);

Reply:
There is definitely a small possibility that a timeout or cancellation request will occur between the time the WriteAsync returns and before we get around to changing the state. But that's OK... the important thing is that we deterministically have only one of the following outcomes:

1. method succeeds and returns to the user (and the stream is NOT disposed)
2. method throws a TimeoutException
3. method throws a OperationCanceledException

In case 1 the stream is guaranteed to NOT be Disposed. In cases 2 and 3 the stream IS Disposed.

Reply:
Note: the window you describe is very small. 

In essence a timeout or cancellation request that occurs during this very small window is treated the same as if it had happened JUST BEFORE WriteAsync completed.

Reply:
Or you can think of it another way... the WriteAsync isn't **really** complete until we ALSO change the state to 2 to record that the WriteAsync completed.

Reply:
I got it. Thanks for the explanation.